File: Arrays/2bestTimeToBuyAndSell/bestTimeToBuyAndSell.java
import java.lang.Math;

public class bestTimeToBuyAndSell {
    private static int maxProfit(int[] prices) {
        int l = 0,
                r = 1,
                maxprofit = 0,
                len = prices.length,
                profit = 0;
        while (r < len) {
            if (prices[l] < prices[r]) {
                profit = prices[r] - prices[l];
                maxprofit = Math.max(maxprofit, profit);
            } else {
                l = r;
            }
            r += 1;
        }
        return maxprofit;
    }

    public static void main(String[] args) {
        int[] nums = { 7, 1, 5, 3, 6, 4 };
        System.out.println(maxProfit(nums));
    }
}

File: Arrays/2bestTimeToBuyAndSell/bestTimeToBuyAndSell.py
# Best time to buy and sell stock - Google interview question (Leet code 121)
# two pointer concept l,r, l=buy, r = sell 7 > 1 1 - 7 -6 if l is greater than r then move l to r and r to +1 index
# now check again l = 1, r=5, maxprofit=4 update max profit, r = 3 maxprofit = 2no update, r = 6 maxprofit=5 update, r=4 maxprofit=5 no update
# found maxprofit space complexity: O(1), time Complexity : O(n)

class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        l, r = 0, 1
        maxProfit = 0
        while r < len(prices):
            if prices[l] < prices[r]:
                profit = prices[r] - prices[l]
                maxProfit = max(maxProfit, profit)
            else:
                l = r
            r += 1
        return maxProfit

print(Solution().maxProfit([7,1,5,3,6,4]))

